fix: replace metrics callback with background scraping to prevent tim…

keivenchang opened the pull request:
#### Overview:

Fix metrics collection timeouts by replacing the synchronous metrics callback with a background scraping task, and resolve various test failures in the metrics system.

#### Details:

- Replace metrics callback with background scraping task to prevent timeouts
- Fix IntGaugeVec cardinality mismatch in Prometheus metrics tests  
- Change metric name from dynamo_component_dynamo_uptime_seconds to dynamo_component_uptime_seconds
- Update test expectations to match actual Prometheus output format
- Fix system_metrics integration test method name from system_status_info to system_status_server_info

#### Where should the reviewer start?

- lib/runtime/src/distributed.rs - Review the removed duplicate metric registration check
- lib/runtime/src/component.rs - Look at the new start_scraping_metrics method
- lib/runtime/src/metrics.rs - Check the fixed test expectations and IntGaugeVec usage

#### Related Issues:

Closes DIS-445

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* New Features
  * Metrics are now scraped continuously in the background for more timely and resilient updates.
  * Prometheus outputs include automatic labels (namespace, component, endpoint) and expanded metric series.
* Refactor
  * Metric registration now permits multiple series sharing a name when labels differ.
  * Uptime metric renamed to uptime_seconds for consistent Prometheus naming.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
fix: replace metrics callback with background scraping to prevent timeouts

- Replace metrics callback with background scraping task to prevent timeouts
- Fix IntGaugeVec cardinality mismatch in Prometheus metrics tests
- Change metric name from dynamo_component_dynamo_uptime_seconds to dynamo_component_uptime_seconds
- Update test expectations to match actual Prometheus output format
- Fix system_metrics integration test method name from system_status_info to system_status_server_info

rmccorm4 commented on lib/runtime/src/component.rs:
```diff
@@ -268,9 +269,11 @@ impl Component {
+    /// Starts a background task that scrapes stats every 5 seconds and updates metrics.
+    /// The callback will simply use the latest scraped data.
+    pub fn start_scraping_metrics(&self) -> Result<()> {
+        const MAX_DELAY: std::time::Duration = std::time::Duration::from_millis(4700);
```
Can you describe or give a name/const to the 4700 here? I'm guessing it's something like 5000ms (5 seconds) minus the 300ms later on in the code - but this could be made clearer with some consts

rmccorm4 commented on lib/runtime/src/metrics.rs:
```diff
@@ -1226,37 +1220,42 @@ dynamo_component_testintcounter{{dynamo_namespace="ns345"}} 12345"#
 
-        let expected_drt_output = format!(
-            r#"# HELP dynamo_component_testcounter A test counter
+        let expected_drt_output = r#"# HELP dynamo_component_testcounter A test counter
```
Do we need a new test case that would've caught the blocking/failing behavior before this fix?

nnshah1 commented on lib/runtime/src/component.rs:
```diff
@@ -281,60 +284,47 @@ impl Component {
-            let (tx, rx) = std::sync::mpsc::channel::<anyhow::Result<()>>();
+        // Use std::thread for the background task to avoid runtime context issues
+        std::thread::spawn(move || {
+            // Create a new runtime for this background thread
```
wanted to double check here - my understanding is that we have primary and secondary runtimes already created - and not sure if we would want to use those or not here - or if we need a new runtime.

@ryanolson , @grahamking - for suggestion here -

keivenchang committed:
fix: improve integration test reliability and performance

- Remove unnecessary test delays (1000ms, 100ms) improving performance ~14x
- Enhance system status server test helpers with proper runtime management
- Use DRT's built-in system_status_server_info() instead of manual spawning
- Add 200/200 soak test to hit /health endpoint for reliability validation
- All 124 integration tests now passing
